lib/puppet/gettext_config.rb

```diff
@@ -0,0 +1,85 @@
+module Puppet::GettextConfig
+  LOCAL_PATH = File.absolute_path('../../locales', File.dirname(__FILE__))
+  POSIX_PATH = File.absolute_path('../../../../share/locale', File.dirname(__FILE__))
+  WINDOWS_PATH = File.absolute_path('../../../../../../puppet/share/locale', File.dirname(__FILE__))
+
+  # Search for puppet gettext config files
+  # @return path to the config, or nil if not found
+  def self.puppet_locale_path
+    if File.exist?(LOCAL_PATH)
+      return LOCAL_PATH
+    elsif Puppet::Util::Platform.windows? && File.exist?(WINDOWS_PATH)
+      return WINDOWS_PATH
+    elsif !Puppet::Util::Platform.windows? && File.exist?(POSIX_PATH)
+      return POSIX_PATH
+    else
+      nil
+    end
+  end
+
+  # Determine which translation file format to use
+  # @param conf_path the path to the gettext config file
+  # @return :mo if in a package structure, :po otherwise
+  def self.translation_mode(conf_path)
+    if WINDOWS_PATH == conf_path || POSIX_PATH == conf_path
+      return :mo
+    else
+      return :po
+    end
+  end
+
+  # If no file path is specified, we search possible locations for
+  # Puppet's gettext config files. If using this module to initialize
+  # gettext for something other than Puppet (e.g. a module), the path
+  # to the config file must be specified.
+  #
+  # If no translation mode is specified, we will use PO files unless
+  # we detect that we are in a Puppet Agent installation, which has
+  # compiled MO translation files.
+  def self.initialize(conf_file_location = nil, mode = nil)
+    begin
+      require 'gettext-setup'
+      require 'locale'
+
+      # Use supplied path, or attempt to detect config files
+      conf_file_location ||= puppet_locale_path
+      mode ||= translation_mode(conf_file_location)
+
+      if conf_file_location
+        if GettextSetup.method(:initialize).parameters.count == 1
+          # For use with old gettext-setup gem versions, will load PO files only
+          GettextSetup.initialize(conf_file_location)
+        else
+          GettextSetup.initialize(conf_file_location, :file_format => mode)
+        end
+        FastGettext.locale = GettextSetup.negotiate_locale(Locale.current.language)
+      else
+        define_fallbacks!
+      end
+    rescue LoadError
+      define_fallbacks!
+    end
+  end
+
+  # Add stub translation methods (`_` and `n_`) to Object if
+  # not already present. These functions may already be loaded
+  # if gettext was initialize previously, or they were stubbed
+  # before. This makes these functions visible everywhere.
+  def self.define_fallbacks!
+    # Don't need to define these twice
+    return if Object.public_methods.include?(:_)
+
+    Object.send(:include, GettextFallbackMethods)
```

Comment:
What do you think about defining the methods top level in their own file, and requiring it in `lib/puppet.rb` when the initialization of Gettext fails? (See newest commit) This will be the same as the stubbing we were doing previously, except with cleaner conditional logic. From what I read [here](https://stackoverflow.com/a/18965660), that should only add the methods to `main`, not include them in the base Object type. If this seems better to you, I may actually open a PR against the gettext-setup gem to do the same thing there, instead of including in Object.